### ISSUE TITLE
feature: Allow ordering service deployments

### ISSUE THREAD

**Issue by vittoriom94** [aws/copilot-cli#2935]
Hello,
when I try to use the attribute "depends_on" in the manifest of a service, I get the error "container dependency does not exist" when I try to generate the cloudformation stack.

The app I generated with copilot has a nginx service that I want to start after its backend services in the copilot generated pipeline.

The structure of the "copilot" folder is:
![image](https://user-images.githubusercontent.com/4594052/137916138-c1ade396-3438-4bb7-9ed1-c911907269f4.png)

My nginx manifest has the following configuration:

```
name: nginx
type: Load Balanced Web Service

http:
  path: '/'

image:
  build
  build: nginx/Dockerfile
  port: 80
  depends_on:
    sport: START

cpu: 256  
memory: 512   
count: 1       
exec: true 

environments:
  test:
    variables:
      DEFAULT_URL: "default.test.cnn.local"
      ECONOMY_URL: "economy.test.cnn.local"
      ENTERTAINMENT_URL: "entertainment.test.cnn.local"
      LEISURE_URL: "leisure.test.cnn.local"
      LIFESTYLE_URL: "lifestyle.test.cnn.local"
      SCIENCE_URL: "science.test.cnn.local"
      SPORT_URL: "sport.test.cnn.local"
```

And this is the beginning of the sport service manifest:
```
# Your service name will be used in naming your resources like log groups, ECS services, etc.
name: sport
type: Backend Service
```

I tried the command to generate the cloudformation stack and the error is:

> copilot svc package -n nginx
> Only found one environment, defaulting to: test
> ✘ generate stack template: convert the container dependency for service nginx: container dependency does not exist

When I try to deploy the service through the pipeline, the error is the same:

> [Container] 2021/10/13 15:46:22 Running command for env in $pl_envs; do
>   for svc in $svcs; do
>   ./copilot-linux svc package -n $svc -e $env --output-dir './infrastructure' --tag $tag;
>   done;
>   for job in $jobs; do
>   ./copilot-linux job package -n $job -e $env --output-dir './infrastructure' --tag $tag;
>   done;
> done;
> 923	✘ generate stack template: convert the container dependency for service nginx: container dependency does not exist
> 

And the outcome is an empty nginx stack.yml and params.json:

> [Container] 2021/10/13 15:46:30 Running command ls -lah ./infrastructure
> total 168K
> drwxr-xr-x 2 root root 4.0K Oct 13 15:46 .
> drwxr-xr-x 9 root root 4.0K Oct 13 15:46 ..
> -rw-r--r-- 1 root root  540 Oct 13 15:46 climate-test.params.json
> -rw-r--r-- 1 root root  15K Oct 13 15:46 climate-test.stack.yml
> -rw-r--r-- 1 root root  540 Oct 13 15:46 default-test.params.json
> -rw-r--r-- 1 root root  15K Oct 13 15:46 default-test.stack.yml
> -rw-r--r-- 1 root root  540 Oct 13 15:46 economy-test.params.json
> -rw-r--r-- 1 root root  15K Oct 13 15:46 economy-test.stack.yml
> -rw-r--r-- 1 root root  558 Oct 13 15:46 entertainment-test.params.json
> -rw-r--r-- 1 root root  15K Oct 13 15:46 entertainment-test.stack.yml
> -rw-r--r-- 1 root root  540 Oct 13 15:46 leisure-test.params.json
> -rw-r--r-- 1 root root  15K Oct 13 15:46 leisure-test.stack.yml
> -rw-r--r-- 1 root root  546 Oct 13 15:46 lifestyle-test.params.json
> -rw-r--r-- 1 root root  15K Oct 13 15:46 lifestyle-test.stack.yml
> -rw-r--r-- 1 root root    0 Oct 13 15:46 nginx-test.params.json
> -rw-r--r-- 1 root root    0 Oct 13 15:46 nginx-test.stack.yml
> -rw-r--r-- 1 root root  540 Oct 13 15:46 science-test.params.json
> -rw-r--r-- 1 root root  15K Oct 13 15:46 science-test.stack.yml
> -rw-r--r-- 1 root root  534 Oct 13 15:46 sport-test.params.json
> -rw-r--r-- 1 root root  15K Oct 13 15:46 sport-test.stack.yml
> 

Am I missing some configuration parameter or do I need to inject the service name in some other way to the manifest?
Thank you!

**Comment by iamhopaul123** [comment]
Hello @vittoriom94. [`depends_on`](https://aws.github.io/copilot-cli/docs/manifest/lb-web-service/#image-depends-on) is used to specify the container dependencies within the same service, and it is not possible to depend on another service (see [ECS doc](https://docs.aws.amazon.com/AmazonECS/latest/APIReference/API_ContainerDefinition.html) as well).

If you want the `nginx` to be deployed after `sport` service, without using the pipeline you could simply deploy these two services in a specific order. However, if using pipeline I think the service deployment dependency would be a feature request for us.

Again could you tell us without service dependency, could you successfully deploy the pipeline? I thought even if you deploy the `nginx` before `sport` it's probably fine because once they all get deployed the pipeline would be working normally? Let me know your thoughts! Thank you!

**Comment by vittoriom94** [comment]
Hi @iamhopaul123, sorry for the delay but I'm doing various tests and every deploy is taking... a bit of time!

I remember I had some failures at the beginning due to nginx not finding the upstream services, but there were also other errors due to my inexperience with the environment, so this time I want to be sure that there is actually an error due to the deploy order.

Thank you for clarifying the use of the depends_on param. The alternative I had was to add an additional step to the pipeline (Source -> Build -> Deploy Backend to Test env -> Deploy Nginx to Test env) and move the nginx deploy to the fourth step, but then I realized that if I have to do some edits to the pipeline and deploy them with the copilot CLI, then the pipeline is reseted to the original one (Source -> Build -> Deploy to Test env).

I haven't seen this possibility in the pipeline.yml documentation, so I guess this would be the feature request you were talking about. 
I'll do some more tests tomorrow to be sure of the outcome of the standard pipeline and let you know the results.
Thank you!

**Comment by iamhopaul123** [comment]
Hello @vittoriom94. Ideally, microservices should be independent with each other. To achieve that, the services should be able to auto retry.

Yes, it would be great for you to let us know the results. Alternatively we could make use of CodePipeline's [runOrder](https://docs.aws.amazon.com/codepipeline/latest/userguide/reference-pipeline-structure.html#action-requirements) and allow Copilot users to create multiple fleets of services to deploy them in a specific order.

**Comment by vittoriom94** [comment]
After trying some more times, it seems that nginx is correctly started without the need of a start order. Sometimes a task o two will fail, due to the backends not being available yet to nginx, but the automatic retry is enough to get the nginx to a steady running state and complete the pipeline successfuly.

If Copilot could handle the runOrder parameter in its pipeline or manifest configuration and forward it to CodePipeline, it would be perfect to avoid some of these tasks to fail!

Thank you!

**Comment by efekarakus** [comment]
Hi @vittoriom94, we've a proposal for this issue here: https://github.com/aws/copilot-cli/issues/3402 !

**Comment by huanjani** [comment]
@vittoriom94: The feature is now released in v1.18.0: https://github.com/aws/copilot-cli/releases/tag/v1.18.0!
